### ISSUE TITLE
Infinite loop

### ISSUE THREAD

**Issue by niklasbuschmann** [ryansolid/solid#125]
The following example leads to an infinite loop between `liftComputation` and `applyUpstreamUpdates`:

```javascript
import { createState, onCleanup } from 'solid-js';
import { render } from 'solid-js/dom';

const App = () => {
  const [state, setState] = createState({a: false, b: false, c: false});

  return <div>
    <button onClick={() => setState({b: true})}>Click me</button>
    <span>{state.a ? 'a' : state.b ? 'b' : state.c ? 'c' : 'fallback'}</span>
  </div>;
}

render(App, document.body);
```

<img width="1392" alt="Screenshot 2020-03-01 at 16 53 33" src="https://user-images.githubusercontent.com/4943215/75628937-ebc9ea80-5bdd-11ea-95d3-e2ca13b28b0e.png">

**Comment by ryansolid** [comment]
Ok, thanks for such a small reproduction case. I'd hit this before with really deep nested recursion patterns (like 10k nested depth) but never something without this heavy nesting. It's probably related to my attempts to optimize ternary operators in the view with the compiler auto wrapping parts of expressions. But this seems like something I can work with. I will look into it. Thanks for reporting.

**Comment by niklasbuschmann** [comment]
Yes, I thought I would share this since it took me quite some time to pin down the issue.

And thank you for creating this - working with solid has been a quite pleasant experience, I really like the simple interface and great performance!

**Comment by ryansolid** [comment]
I fixed the issue but it took some thinking. I've included it below for completeness. In any case fixed in Babel Plugin JSX DOM Expressions 0.16.4. Current version of Babel Preset Solid will support it but might need to blow away the package locks to get it to work.

------------------------------
I was thinking I can reduce the output. But I'm thinking now maybe I can't. The reason I started doing this is to remove recalculation on changing values because it could be expensive.

Ie... `state.count > 2 ? <CompA /> : <CompB />` 

As count becomes 3 and 4 and so on it would keep recreating `CompA`. If I independently wrap `state.count > 2` then it is possible to only retrigger when switching from false to true so each increase in count would not retrigger. The issue of course is it isn't simple enough to look for a boolean operator.

Consider `countBiggerThan2() ? <CompA/> :  <CompB/>`. Without knowing the internals I can't know if I need to guard or not.  Or how about  `state.count ? <CompA/> :  <CompB/>`? There is no way to know this is any different at compile time than `state.show ? <CompA/> :  <CompB/>`. But if count keeps updating after it initially shows A it is very different than a boolean that is still true. This case is super common with like `state.list.length`.

So basically as long as it's dynamic from static analysis I either wrap always to be safe or wrap never and put this on the end user. I think the latter is unacceptable. The only heuristic that I could do is be unoptimized for these `state.count` examples unless the end user makes it clearly comparison grammar (Binary Expression). Ie `state.count` has to be `state.count > 0` or `state.count !== 0` in order to wrap. Of course that wouldn't work with functions so it'd only apply to state expressions.

So instead of worrying about that I realize it was probably prudent to change the definition of expensive work. Before member expressions like proxies were included as I was using a standard dynamic expression check. But now I realize the only potential for expensive work comes from nested components or function calls. `state.a ? 'a' : state.b ? 'b' : state.c ? 'c' : 'fallback'` could never be that expensive. So it doesn't even need to be wrapped piece-wise at all.

And that's where this sits.
